### PR TITLE
fix URL Rewrite issues

### LINKED ISSUE #1598: Method transform not working when combined with URL rewrite
**Do you want to request a *feature* or report a *bug*?**
I wanna report a bug.

**What is the current behavior?**
When I combine method transform + url rewrite, the method transform does not work.

**What is the expected behavior?**
The method transform is supposed to work.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem**
1. Create a GET endpoint that rewrites to a another endpoint (listen to get and map to http://httpbin.org/post).
2. Add method transform from GET to POST.
3. Test - not working - method not allowed.

**Which versions of Tyk affected by this issue? Did this work in previous versions of Tyk?**
Tested in Tyk Cloud (Dashboard v1.5.4) and locally with Tyk Dashboard 1.5.4.

### LINKED ISSUE #2268: UrlRewrite/Method transform doesn't work if caching is enabled
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
If you enable URL Rewrite/Method transform and Cache plugin for an endpoint, URL Rewrite/Method transform doesn't work.

**What is the expected behavior?**
URL Rewrite/Method transform plugin should work

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem**

API Defintion

```json
{
  "api_model": {},
  "api_definition": {
    "api_id": "10a5aa802e0346a653eeb71f8e9d9234",
    "jwt_issued_at_validation_skew": 0,
    "upstream_certificates": {},
    "use_keyless": true,
    "enable_coprocess_auth": false,
    "custom_middleware": {
      "pre": [],
      "post": [],
      "post_key_auth": [],
      "auth_check": {
        "name": "",
        "path": "",
        "require_session": false
      },
      "response": [],
      "driver": "",
      "id_extractor": {
        "extract_from": "",
        "extract_with": "",
        "extractor_config": {}
      }
    },
    "disable_quota": false,
    "custom_middleware_bundle": "",
    "cache_options": {
      "cache_timeout": 60,
      "enable_cache": true,
      "cache_all_safe_requests": false,
      "cache_response_codes": [],
      "enable_upstream_cache_control": false,
      "cache_control_ttl_header": ""
    },
    "enable_ip_blacklisting": false,
    "tag_headers": [],
    "pinned_public_keys": {},
    "expire_analytics_after": 0,
    "domain": "",
    "openid_options": {
      "providers": [],
      "segregate_by_client": false
    },
    "active": true,
    "jwt_expires_at_validation_skew": 0,
    "config_data": {},
    "notifications": {
      "shared_secret": "",
      "oauth_on_keychange_url": ""
    },
    "auth": {
      "auth_header_name": "",
      "signature": {
        "algorithm": "",
        "header": "",
        "secret": "",
        "allowed_clock_skew": 0,
        "error_code": 0,
        "error_message": ""
      },
      "validate_signature": false,
      "use_certificate": false
    },
    "check_host_against_uptime_tests": false,
    "auth_provider": {
      "name": "",
      "storage_engine": "",
      "meta": {}
    },
    "blacklisted_ips": [],
    "hmac_allowed_clock_skew": -1,
    "dont_set_quota_on_create": false,
    "uptime_tests": {
      "check_list": [],
      "config": {
        "expire_utime_after": 0,
        "service_discovery": {
          "use_discovery_service": false,
          "query_endpoint": "",
          "use_nested_query": false,
          "parent_data_path": "",
          "data_path": "",
          "cache_timeout": 60
        },
        "recheck_wait": 0
      }
    },
    "enable_jwt": false,
    "do_not_track": false,
    "name": "test",
    "slug": "test",
    "oauth_meta": {
      "allowed_access_types": [],
      "allowed_authorize_types": [],
      "auth_login_redirect": ""
    },
    "CORS": {
      "enable": false,
      "max_age": 24,
      "allow_credentials": false,
      "exposed_headers": [],
      "allowed_headers": [],
      "options_passthrough": false,
      "debug": false,
      "allowed_origins": [],
      "allowed_methods": []
    },
    "event_handlers": {
      "events": {}
    },
    "proxy": {
      "target_url": "http://httpbin.org/",
      "service_discovery": {
        "endpoint_returns_list": false,
        "cache_timeout": 0,
        "parent_data_path": "",
        "query_endpoint": "",
        "use_discovery_service": false,
        "_sd_show_port_path": false,
        "target_path": "",
        "use_target_list": false,
        "use_nested_query": false,
        "data_path": "",
        "port_data_path": ""
      },
      "check_host_against_uptime_tests": false,
      "transport": {
        "ssl_insecure_skip_verify": false,
        "ssl_ciphers": [],
        "ssl_min_version": 0,
        "proxy_url": ""
      },
      "target_list": [],
      "preserve_host_header": false,
      "strip_listen_path": true,
      "enable_load_balancing": false,
      "listen_path": "/test/",
      "disable_strip_slash": false
    },
    "client_certificates": [],
    "use_basic_auth": false,
    "version_data": {
      "not_versioned": true,
      "default_version": "",
      "versions": {
        "Default": {
          "name": "Default",
          "expires": "",
          "paths": {
            "ignored": [],
            "white_list": [],
            "black_list": []
          },
          "use_extended_paths": true,
          "global_headers": {},
          "global_headers_remove": [],
          "global_size_limit": 0,
          "override_target": "",
          "extended_paths": {
            "cache": [
              "/post"
            ],
            "method_transforms": [
              {
                "path": "/post",
                "method": "GET",
                "to_method": "POST",
                "fromDashboard": true
              }
            ]
          }
        }
      }
    },
    "use_standard_auth": false,
    "session_lifetime": 0,
    "hmac_allowed_algorithms": [],
    "disable_rate_limit": false,
    "definition": {
      "location": "header",
      "key": "x-api-version",
      "strip_path": false
    },
    "use_oauth2": false,
    "jwt_not_before_validation_skew": 0,
    "allowed_ips": [],
    "org_id": "5cd00524d85f245425510a78",
    "enable_ip_whitelisting": false,
    "global_rate_limit": {
      "rate": 0,
      "per": 0
    },
    "enable_context_vars": false,
    "tags": [],
    "basic_auth": {
      "disable_caching": false,
      "cache_ttl": 0,
      "extract_from_body": false,
      "body_user_regexp": "",
      "body_password_regexp": ""
    },
    "session_provider": {
      "name": "",
      "storage_engine": "",
      "meta": {}
    },
    "strip_auth_data": false,
    "id": "5cd00554d85f245425510a79",
    "certificates": [],
    "enable_signature_checking": false,
    "use_openid": false,
    "internal": false,
    "jwt_skip_kid": false,
    "enable_batch_request_support": false,
    "response_processors": [],
    "use_mutual_tls_auth": false
  },
  "hook_references": [],
  "is_site": false,
  "sort_by": 0
}
```

`curl https://tyk-gateway:8181/test/post` will return 405 error

Remove cache plugin from `/post` endpoint and above command will work

**Which versions of Tyk affected by this issue? Did this work in previous versions of Tyk?**
master/release2.8

### LINKED ISSUE #2618: URLRewriteMiddleware modifies the original request
**Branch/Environment/Version**
- Branch/Version: [e.g. Master/Release/Stable/Feature branch]
- Environment: [e.g. On-prem/Hybrid/MDCB]

**Describe the bug**
The `URLRewriteMiddleware` modifies the original request, this is a source of bugs, because the request object is a fact, and if new information needs to be added there `Context` for it.

I came across this while debugging an issue with url rewriting feature.

### PR DESCRIPTION
**Pull request by komalsukhani**
Fixes #2268 
Fixes #1598 
Fixes #2618 

`URLRewrite` middleware changes `r.URL` to the `target` path. This breaks the middleware chain, as each middleware using `r.URL`, to check if it is enabled. So instead of changing `r.URL` directly in the URLRewrite middleware, I moved it to DummyProxyHandler.

Similarly, I also changed `Method Tranform` middleware.

The PR also changes `cache` middleware, moving it at the end of middleware chain and also incorporating above changes.

### CODE REVIEW

**Comment by gernest** [comment]
@furkansenharputlu PTAL

File: gateway/gateway_test.go [hunk 1]
```diff
@@ -1111,6 +1111,62 @@ func TestCacheAllSafeRequests(t *testing.T) {
...
 
+func TestCacheWithAdvanceUrlRewrite(t *testing.T) {
+	ts := StartTest()
+	defer ts.Close()
+	cache := storage.RedisCluster{KeyPrefix: "cache-"}
+	defer cache.DeleteScanMatch("*")
+
+	BuildAndLoadAPI(func(spec *APISpec) {
+		version := spec.VersionData.Versions["v1"]
+		json.Unmarshal([]byte(`{
```
Comment (furkansenharputlu) [inline gateway/gateway_test.go]: Please don't use json for API configuration like the other test you added.

File: gateway/api.go [hunk 2]
```diff
@@ -2087,6 +2087,20 @@ func ctxGetOrigRequestURL(r *http.Request) *url.URL {
 	return nil
 }
 
+func ctxSetUrlRewriteTarget(r *http.Request, url *url.URL) {
```
Comment (furkansenharputlu) [inline gateway/api.go]: `ctxSetURLRewriteTarget` same for other new added `Url`s(not existing ones) throughout the PR.

File: gateway/api_loader.go [hunk 3]
```diff
@@ -422,7 +421,9 @@ func processSpec(spec *APISpec, apisByListen map[string]int,
 			chainArray = append(chainArray, createDynamicMiddleware(obj.Name, false, obj.RequireSession, baseMid))
 		}
 	}
-
+	//Do not add middlewares after cache middleware.
+	//It will not get executed
+	mwAppendEnabled(&chainArray, &RedisCacheMiddleware{BaseMiddleware: baseMid, CacheStore: &cacheStore})
```
Comment (furkansenharputlu) [inline gateway/api_loader.go]: AFAIK, `VirtualEndpoint` was working. Is there a specific case or something?

File: gateway/gateway_test.go [hunk 4]
```diff
@@ -1111,6 +1111,62 @@ func TestCacheAllSafeRequests(t *testing.T) {
...
+		spec.Proxy.ListenPath = "/"
+		spec.VersionData.Versions["v1"] = version
+	})
+
+	headerCache := map[string]string{"x-tyk-cached-response": "1"}
+	matchHeaders := map[string]string{"rewritePath": "newpath"}
+	randomheaders := map[string]string{"something": "abcd"}
+
+	ts.Run(t, []test.TestCase{
+		{Method: "GET", Path: "/test", Headers: matchHeaders, HeadersNotMatch: headerCache, Delay: 10 * time.Millisecond},
```
Comment (furkansenharputlu) [inline gateway/gateway_test.go]: Please use `http.MethodGet, http.MethodPost` throughout the PR.

File: gateway/mw_url_rewrite.go [hunk 5]
```diff
@@ -384,7 +386,9 @@ func (m *URLRewriteMiddleware) ProcessRequest(w http.ResponseWriter, r *http.Req
 	if err != nil {
 		log.Error("URL Rewrite failed, could not parse: ", p)
 	} else {
-		r.URL = newURL
+		//Setting new path here breaks request middleware
```
Comment (furkansenharputlu) [inline gateway/mw_url_rewrite.go]: No need to `else` statement

File: ctx/ctx.go [hunk 6]
```diff
@@ -8,8 +8,10 @@ import (
 	"github.com/TykTechnologies/tyk/user"
 )
 
+type Key uint
```
Comment (furkansenharputlu) [inline ctx/ctx.go]: More descriptive name like `ContextKey`

Comment (gernest) [reply ctx/ctx.go]: The package is `ctx` so it looks descriptive enough since you refer to it as `ctx.Key`

Comment (furkansenharputlu) [reply ctx/ctx.go]: I didn't realize it is in `ctx` package, ok.

File: gateway/mw_url_rewrite.go (hunk 5 above)
Comment (gernest) [reply gateway/mw_url_rewrite.go]: We don't call `ctxSetURLRewriteTarget` when `err!=nil` and we don't return inside `err` check, else is required here.

File: gateway/api_loader.go (hunk 3 above)
Comment (gernest) [reply gateway/api_loader.go]: I have no idea about `VirtualEndpoint` you are mentioning, however this make sense to me. As it ensures caching is the last middleware in the chain.

File: gateway/mw_url_rewrite.go (hunk 5 above)
Comment (furkansenharputlu) [reply gateway/mw_url_rewrite.go]: yep I missed.

**Comment by buger** [comment]
Does it also fix the case when Body transform does not work when URL rewrite enabled?

**Pull request merged**